src/ffi/buffer.rs

```diff
@@ -0,0 +1,242 @@
+    /// Returns the buffer as a Rust slice.
+    pub fn as_slice(&self) -> &[T] {
+        let Inner { len, buf } = &self.inner;
+
+        // SAFETY: `AscBuf` can only be constructed from an `AscBuffer` which
+        // correctly allocates the storage starting at `buf` to have `len`
+        // elements. Additionally we *assume* that all `AscBuf` references from
+        // host calls are valid in the same way.
+        unsafe { slice::from_raw_parts((buf as *const A).cast(), *len) }
+    }
+}
+
+impl<T, A> Debug for AscBuf<T, A>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        Debug::fmt(&self.as_slice(), f)
+    }
+}
+
+/// An owned AssemblyScript dynamically sized buffer with elements of type `T`
+/// and aligned to `Alignment`.
+#[repr(transparent)]
+pub struct AscBuffer<T, Alignment = T> {
+    _type: PhantomData<T>,
+    inner: Inner<[Alignment]>,
+}
+
+impl<T, A> AscBuffer<T, A> {
+    /// Creates a new AssemblyScript buffer from the specified slice.
+    pub fn new(slice: impl AsRef<[T]>) -> Box<Self> {
+        let slice = slice.as_ref();
+
+        // SAFETY: The allocated buffer is guaranteed to be completely
+        // initialized.
+        unsafe {
+            let mut buffer = alloc_buffer::<T, A>(slice.len());
+            buffer.inner.len = slice.len();
+
+            // NOTE: Use `ptr::copy` here since the allocated buffer contains
+            // unintialized memory. It is considered undefined behaviour to
+            // create a reference to uninitialized memory in Rust.
+            ptr::copy(
+                slice.as_ptr(),
+                buffer.inner.buf.as_mut_ptr().cast(),
+                slice.len(),
+            );
+
+            buffer
+        }
+    }
+
+    /// Returns a reference to a borrowed AssemblyScript buffer.
+    pub fn as_buf(&self) -> &AscBuf<T, A> {
+        unsafe { &*(&self.inner.len as *const usize).cast::<AscBuf<T, A>>() }
+    }
+
+    /// Returns an FFI-safe pointer to an AssemblyScript buffer.
+    pub fn as_buf_ptr(&self) -> *const AscBuf<T, A> {
+        self.as_buf() as *const _
+    }
+}
+
+impl<T, A> Deref for AscBuffer<T, A> {
+    type Target = AscBuf<T, A>;
+
+    fn deref(&self) -> &Self::Target {
+        self.as_buf()
+    }
+}
+
+impl<T, A> Debug for AscBuffer<T, A>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        Debug::fmt(self.as_buf(), f)
+    }
+}
+
+/// Returns the memory layout for an AssemblyScript buffer with the specified
+/// dynamic length.
+fn buffer_layout<T, A>(len: usize) -> Result<Layout, LayoutErr> {
+    let (layout, _) = Layout::new::<AscBuf<T, A>>().extend(Layout::array::<T>(len)?)?;
+    // NOTE: Pad to alignment for C ABI compatibility. See
+    // <https://doc.rust-lang.org/std/alloc/struct.Layout.html#method.extend>
+    Ok(layout.pad_to_align())
+}
+
+/// A Rust dynamically sized type fat pointer.
+struct DstRef {
+    #[allow(dead_code)]
+    ptr: *const u8,
```

Comment:
Is `u8` the right size for a pointer? Shouldn't we use `usize` ?

Reply:
This is just saying that `ptr` is some pointer to `u8`s. If we use `usize` then the compiler might make some assumptions on the alignment of the pointer which would be incorrect.

To be perfectly honest, I'm not 100% sure about the correctness here :stuck_out_tongue:. It was mostly picked because it is the return type of the global allocator function: https://doc.rust-lang.org/std/alloc/fn.alloc.html